modules/programs/vscode/haskell.nix

```diff
@@ -1,62 +1,33 @@
 { pkgs, config, lib, ... }:
 
-with lib;
+let cfg = config.programs.vscode.haskell;
 
-let
-
-  cfg = config.programs.vscode.haskell;
-
-  defaultHieNixExe = hie-nix.hies + "/bin/hie-wrapper";
-  defaultHieNixExeText =
-    literalExpression ''"''${pkgs.hie-nix.hies}/bin/hie-wrapper"'';
-
-  hie-nix = pkgs.hie-nix or (abort ''
-    vscode.haskell: pkgs.hie-nix missing. Please add an overlay such as:
-    ${exampleOverlay}
-  '');
-
-  exampleOverlay = ''
-    nixpkgs.overlays = [
-      (self: super: { hie-nix = import ~/src/hie-nix {}; })
-    ]
-  '';
-
-in {
+in with lib; {
   options.programs.vscode.haskell = {
     enable = mkEnableOption "Haskell integration for Visual Studio Code";
 
-    hie.enable = mkOption {
-      type = types.bool;
-      default = true;
-      description = "Whether to enable Haskell IDE engine integration.";
-    };
+    useGhcup = mkEnableOption "using GHCup for the Haskell integration";
 
-    hie.executablePath = mkOption {
-      type = types.path;
-      default = defaultHieNixExe;
-      defaultText = defaultHieNixExeText;
-      description = ''
-        The path to the Haskell IDE Engine executable.
-        </para><para>
-        Because hie-nix is not packaged in Nixpkgs, you need to add it as an
-        overlay or set this option. Example overlay configuration:
-        <programlisting language="nix">${exampleOverlay}</programlisting>
-      '';
-      example = literalExpression ''
-        (import ~/src/haskell-ide-engine {}).hies + "/bin/hie-wrapper";
-      '';
+    hlsPackage = mkPackageOption pkgs "Haskell language server" {
+      default = [ "haskell-language-server" ];
     };
+
+    ghcupPackage =
+      mkPackageOption pkgs "GHCup" { default = [ "haskellPackages" "ghcup" ]; };
   };
 
   config = mkIf cfg.enable {
-    programs.vscode.userSettings = mkIf cfg.hie.enable {
-      "languageServerHaskell.enableHIE" = true;
-      "languageServerHaskell.hieExecutablePath" = cfg.hie.executablePath;
-    };
-
-    programs.vscode.extensions =
-      [ pkgs.vscode-extensions.justusadam.language-haskell ]
-      ++ lib.optional cfg.hie.enable
-      pkgs.vscode-extensions.alanz.vscode-hie-server;
+    programs.vscode.userSettings = mkMerge [
+      (mkIf cfg.useGhcup {
+        "haskell.ghcupExecutablePath" = cfg.ghcupPackage;
+        "haskell.manageHLS" = "GHCup";
+      })
+
+      (mkIf (!cfg.useGhcup) {
+        "haskell.serverExecutablePath" = cfg.hlsPackage;
+      })
+    ];
+
+    programs.vscode.extensions = [ pkgs.vscode-extensions.haskell.haskell ];
```

Comment:
This should probably still include `justusadam.language-haskell`, which is a dependency of `haskell.haskell`. Otherwise users will have to install it impurely.

Reply:
Thanks, didn’t know

Reply:
Has been done